### ISSUE TITLE
This plugin will no longer be needed in Caddy v2.7.0

### ISSUE THREAD

**Issue by francislavoie** [kirsch33/realip#14]
Just wanted to mention here that this plugin will no longer be needed because of the work done in https://github.com/caddyserver/caddy/pull/5104.

The real client IP will be parsed if `trusted_proxies` is configured in global options, and the header it's read from can be configured with `client_ip_headers` (defaulting to `X-Forwarded-For`). The logs will include it as a new `client_ip` field.

Compared to this plugin, this approach doesn't affect `req.RemoteAddr`, that's left intact, but any other modules/handlers that might care to get the real client IP can get it from the request context.

**Comment by lxhao61** [comment]
When will Caddy v2.7.0 be released?

**Comment by francislavoie** [comment]
Caddy doesn't follow a release schedule. A release will be made when we have enough to release.

**Comment by kirsch33** [comment]
thanks for the heads up. i have referenced this discussion on the main page for this repo.

**Comment by crapthings** [comment]
can i use this for 2.6.4?

how to install this module?

**Comment by francislavoie** [comment]
No @crapthings. You could build from the master branch though. It's not a plugin, it's changes in Caddy itself.

**Comment by mgagnant** [comment]
Hello,
i use caddy 2.7 beta for testing the recuperation of source IP.
I have the field "client_ip" but the ip is the container ip....

where i must put trusted_proxies parameters please ?

after in my PHP code i use : 
$user_ip = $_SERVER['HTTP_X_FORWARDED_FOR'];
echo "IP HTTP_X_FORWARDED_FOR : ".$user_ip."<br>";

Thanks for your reply.

**Comment by francislavoie** [comment]
In global options, at the top of your Caddyfile:

```
{
	servers {
		trusted_proxies static <ips...>
	}
}
```

**Comment by mgagnant** [comment]
Thanks a lot for your Quick answer.
Unfortunately still the Same problem. The client_ip value is 10.0.0.3 which is the container IP. 
maybe i made a mistake on trusted_proxies.
What are the statics ips I must enter ? I try with 10.0.0.3 .
Thanks again

**Comment by francislavoie** [comment]
If you're in Docker then you might be running into the problem where Docker is using a userland proxy which causes all TCP packets to appear that they're coming from Docker itself. That's not something you can fix with Caddy itself, it's a problem with Docker. See https://github.com/moby/moby/issues/15086

**Comment by mgagnant** [comment]
Ok thanks again for your quick answer. I will investigate on docker.

**Comment by thesantiagosp** [comment]
I'm using the new sintax but i don't know where I'm failing. I want to pass the IP present in the cloudflare header but it's not working as i thought. For example, with this config:

```
{
	servers {
		trusted_proxies static cloudflare_ips_range
		client_ip_headers CF-Connecting-IP
	}
}
domain.example {
	reverse_proxy example {
		header_up X-Forwarded-For {client_ip}
	}
}
```

The application is not getting the header. If i change the trusted header to:
`client_ip_headers X-Forwarded-For`

Then it works. Why is not taking the IP from the CF header? Did I misunderstood something?

**Comment by francislavoie** [comment]
`cloudflare_ips_range` is not a thing.

If you want to automatically get the trusted proxies IPs for cloudflare, use this plugin https://github.com/WeidiDeng/caddy-cloudflare-ip instead of `static`.

Also remove your `header_up X-Forwarded-For` line, that's incorrect; Caddy will automatically set that header correctly if `trusted_proxies` is correct.

**Comment by thesantiagosp** [comment]
`cloudflare_ips_range` was just and example to not write all the IPs.

The problem is that some applications does not properly read the header if it contains multiple IP address `client_ip, cloudflare_ip`. That's why I want to overwrite the header and just set the client IP without the intermediate.

**Comment by francislavoie** [comment]
> The problem is that some applications does not properly read the header if it contains multiple IP address 

Then those should be fixed. They're not conformant to spec otherwise.

Also you might need to use `Cf-Connecting-Ip` and not `CF-Connecting-IP` (lowercase `f` and `p`)

**Comment by felipecrs** [comment]
[Cloudflare's instructions for Caddy](https://developers.cloudflare.com/support/troubleshooting/restoring-visitor-ips/restoring-original-visitor-ips/#caddy) points to use `header_up` instead.

Would it be fair to say Cloudflare's instructions are outdated/suboptimal?

**Comment by francislavoie** [comment]
Yes, Cloudflare's instructions are very wrong.

**Comment by felipecrs** [comment]
Would you please help me building the correct configuration?

```
{
	servers {
		trusted_proxies static 127.0.0.1
		client_ip_headers CF-Connecting-IP
	}
}
```

Doesn't seem to be working for me too (consider 127.0.0.1 as if it was Cloudflare's IP).

**Comment by francislavoie** [comment]
127.0.0.1 is definitely not Cloudflare. Show how you're testing this. Show your full config. Show logs.

**Comment by felipecrs** [comment]
In my case I'm using Cloudflared running in the same machine as Caddy itself. All connections pass through Cloudflared in this case, therefore 127.0.0.1.

The thing is, with Nginx, this works:

```
set_real_ip_from 127.0.0.1;
real_ip_header CF-Connecting-IP;
```

- https://github.com/brenner-tobias/addon-cloudflared/pull/852

But I'm [migrating to Caddy](https://github.com/brenner-tobias/addon-cloudflared/pull/843), and:

```
{
	servers {
		trusted_proxies static 127.0.0.1
		client_ip_headers CF-Connecting-IP
	}
}
```

Does not do the trick. The real IP continues to be 127.0.0.1.

**Comment by francislavoie** [comment]
To be clear, `RemoteAddr` on the request will not change, it's not meant to change, and should never change. This plugin was a hack. You should be reading `X-Forwarded-For` in your upstream app, which is what `trusted_proxies` allows you to safely do.

**Comment by felipecrs** [comment]
I.e. there is no [Nginx realip](https://nginx.org/en/docs/http/ngx_http_realip_module.html) equivalent in Caddy?

**Comment by francislavoie** [comment]
No. Why do you need that? Use `client_ip` in your logs, use `X-Forwarded-For` in your upstream app. That way there's no data loss, the original IP is still available, and the "real IP" is available via standard means.

**Comment by felipecrs** [comment]
I don't control my app by the way, which is Home Assistant in this case.

**Comment by francislavoie** [comment]
You do. https://www.home-assistant.io/integrations/http/ Enable `use_x_forwarded_for` and `trusted_proxies`. That's it.

**Comment by felipecrs** [comment]
Oh ok. And I already have it enabled. The issue must be something else. Maybe `X-Forwarded-For` is not getting rewritten after all. I will do more testing.

**Comment by francislavoie** [comment]
Did you set up `trusted_proxies` in HA? you need that too, it ignores `X-Forwarded-For` otherwise

**Comment by felipecrs** [comment]
Yes, I do. `trusted_proxies` in HA is configured with the subnet of my docker internal network where cloudflared + caddy container is running (it is important to remind that it works with cloudflared + nginx and the [snippet I mentioned above](https://github.com/kirsch33/realip/issues/14#issuecomment-2963655093)).

**Comment by felipecrs** [comment]
@francislavoie I truly appreciate the time you took to answer my questions.

I finished my additional testing, and found out that:

```caddyfile
{
	servers {
		trusted_proxies static 127.0.0.1
		client_ip_headers CF-Connecting-IP
	}
}
ha.example.com {
    reverse_proxy homeassistant:8123
}
```

Indeed **does not work**.

But [Cloudflare's suggestion](https://developers.cloudflare.com/support/troubleshooting/restoring-visitor-ips/restoring-original-visitor-ips/#caddy):

```caddyfile
ha.example.com {
    @cloudflared remote_ip 127.0.0.1

    reverse_proxy @cloudflared homeassistant:8123 {
        header_up X-Forwarded-For {http.request.header.CF-Connecting-IP}
    }

    reverse_proxy homeassistant:8123
}
```

**Does work**.

It's a shame because the Caddyfile becomes a lot more verbose:

- https://github.com/brenner-tobias/addon-cloudflared/pull/843/commits/159cb8e2fe013fa3c5ef706b7a78ea2ee0ea66f1

But at least it's no longer a blocker for migrating to Caddy.

---

I don't know why it is not working with the first case. If it helps, I captured a debug log for the request:

```json
{
  "level": "debug",
  "ts": 1749688879.7313411,
  "logger": "http.handlers.reverse_proxy",
  "msg": "upstream roundtrip",
  "upstream": "homeassistant:8123",
  "duration": 0.249233582,
  "request": {
    "remote_ip": "127.0.0.1",
    "remote_port": "43770",
    "client_ip": "185.153.176.160",
    "proto": "HTTP/1.1",
    "method": "POST",
    "host": "ha.example.com",
    "uri": "/auth/login_flow/955249d8aed3711909f0ad0040daab69",
    "headers": {
      "Content-Type": ["text/plain;charset=UTF-8"],
      "Via": ["1.1 Caddy"],
      "Content-Length": ["81"],
      "Sec-Fetch-Mode": ["cors"],
      "X-Forwarded-Proto": ["https"],
      "Accept-Language": ["pt-BR,pt;q=0.9"],
      "Origin": ["https://ha.example.com"],
      "Sec-Ch-Ua-Mobile": ["?1"],
      "Sec-Fetch-Site": ["same-origin"],
      "Cf-Visitor": ["{\"scheme\":\"https\"}"],
      "Sec-Ch-Ua": [
        "\"Google Chrome\";v=\"137\", \"Chromium\";v=\"137\", \"Not/A)Brand\";v=\"24\""
      ],
      "Accept-Encoding": ["gzip, br"],
      "Accept": ["*/*"],
      "Cdn-Loop": ["cloudflare; loops=1"],
      "X-Forwarded-Host": ["ha.example.com"],
      "Cookie": ["REDACTED"],
      "Cf-Warp-Tag-Id": ["f80f0775-dbba-466d-9045-2064a8b9ac5b"],
      "Priority": ["u=1, i"],
      "User-Agent": [
        "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Mobile Safari/537.36"
      ],
      "X-Forwarded-For": ["185.153.176.160, 127.0.0.1"],
      "Cf-Connecting-Ip": ["185.153.176.160"],
      "Cf-Ipcountry": ["BR"],
      "Cf-Ray": ["94e54d486888e9c5-GRU"],
      "Sec-Ch-Ua-Platform": ["\"Android\""],
      "Sec-Fetch-Dest": ["empty"]
    },
    "tls": {
      "resumed": false,
      "version": 772,
      "cipher_suite": 4865,
      "proto": "",
      "server_name": "ha.example.com.localhost"
    }
  },
  "headers": {
    "Referrer-Policy": ["no-referrer"],
    "X-Content-Type-Options": ["nosniff"],
    "Server": [""],
    "X-Frame-Options": ["SAMEORIGIN"],
    "Content-Length": ["222"],
    "Content-Encoding": ["gzip"],
    "Date": ["Thu, 12 Jun 2025 00:41:19 GMT"],
    "Content-Type": ["application/json"]
  },
  "status": 200
}
```

**Comment by francislavoie** [comment]
Enable `debug` global option to get logs from reverse_proxy and see what headers are sent down.

The log shows that client_ip does indeed get filled correctly with the real IP, so trusted_proxies did what is supposed to.

**Comment by felipecrs** [comment]
```caddyfile
{
        debug
        servers {
                trusted_proxies static 127.0.0.1
                client_ip_headers CF-Connecting-IP
        }
}

# Allows to wait for Caddy to be ready before starting Cloudflared
https://healthcheck.localhost {
        tls internal
        respond "OK" 200
}

ha.example.com {
        reverse_proxy http://homeassistant:8123
}

traccar-ha.example.com {
        reverse_proxy http://a0d7b954-traccar:8082
}

go2rtc-ha.example.com {
        # Block connections from Cloudflared as service is internal only
        @cloudflared remote_ip 127.0.0.1
        handle @cloudflared {
                respond "This service can only be accessed from local network." 403
        }

        reverse_proxy http://ccab4aaf-frigate-beta:1984
}

# Catch-all service for any unmatched requests
:80 {
        respond "This service was not found." 404
}

# Only used during automatic Let's Encrypt certificate generation
https://ha.example.com.localhost https://traccar-ha.example.com.localhost https://go2rtc-ha.example.com.localhost {
        tls internal
        respond 407
}
```

If I isolate the logs, these are the only ones I get upon a request:

First:

```json
{
  "level": "debug",
  "ts": 1749694627.577186,
  "logger": "http.handlers.reverse_proxy",
  "msg": "selected upstream",
  "dial": "homeassistant:8123",
  "total_upstreams": 1
}
```

Second:

```json
{
  "level": "debug",
  "ts": 1749694627.8309836,
  "logger": "http.handlers.reverse_proxy",
  "msg": "upstream roundtrip",
  "upstream": "homeassistant:8123",
  "duration": 0.253721579,
  "request": {
    "remote_ip": "127.0.0.1",
    "remote_port": "46170",
    "client_ip": "185.153.176.160",
    "proto": "HTTP/1.1",
    "method": "POST",
    "host": "ha.example.com",
    "uri": "/auth/login_flow/8a93f22b355dda48798f5f249350b321",
    "headers": {
      "Cf-Ipcountry": ["BR"],
      "Cookie": ["REDACTED"],
      "Cf-Warp-Tag-Id": ["162237da-e721-42f6-8714-9d20daeb2e44"],
      "Sec-Fetch-Dest": ["empty"],
      "Cf-Visitor": ["{\"scheme\":\"https\"}"],
      "Sec-Ch-Ua": [
        "\"Google Chrome\";v=\"137\", \"Chromium\";v=\"137\", \"Not/A)Brand\";v=\"24\""
      ],
      "Accept-Encoding": ["gzip, br"],
      "Sec-Ch-Ua-Mobile": ["?1"],
      "X-Forwarded-Host": ["ha.example.com"],
      "Priority": ["u=1, i"],
      "Sec-Ch-Ua-Platform": ["\"Android\""],
      "Sec-Fetch-Site": ["same-origin"],
      "Origin": ["https://ha.example.com"],
      "Content-Type": ["text/plain;charset=UTF-8"],
      "Cf-Connecting-Ip": ["185.153.176.160"],
      "Accept": ["*/*"],
      "Sec-Fetch-Mode": ["cors"],
      "Via": ["1.1 Caddy"],
      "User-Agent": [
        "Mozilla/5.0 (Linux; Android 10; K) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/137.0.0.0 Mobile Safari/537.36"
      ],
      "X-Forwarded-For": ["185.153.176.160, 127.0.0.1"],
      "X-Forwarded-Proto": ["https"],
      "Content-Length": ["82"],
      "Accept-Language": ["pt-BR,pt;q=0.9"],
      "Cf-Ray": ["94e5d99df94800c5-GRU"],
      "Cdn-Loop": ["cloudflare; loops=1"]
    },
    "tls": {
      "resumed": false,
      "version": 772,
      "cipher_suite": 4865,
      "proto": "",
      "server_name": "ha.example.com.localhost"
    }
  },
  "headers": {
    "X-Frame-Options": ["SAMEORIGIN"],
    "Content-Length": ["223"],
    "Content-Encoding": ["gzip"],
    "Date": ["Thu, 12 Jun 2025 02:17:07 GMT"],
    "Content-Type": ["application/json"],
    "Referrer-Policy": ["no-referrer"],
    "X-Content-Type-Options": ["nosniff"],
    "Server": [""]
  },
  "status": 200
}
```

**Comment by francislavoie** [comment]
`"X-Forwarded-For": ["185.153.176.160, 127.0.0.1"],`

Yep, it's passed down just fine. Your problem must be your HA configuration. You need `127.0.0.1` as part of `trusted_proxies` in your HA config.

**Comment by felipecrs** [comment]
This is my HA configuration by the way:

```yaml
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - 172.30.33.0/24
```

Maybe `X-Forwarded-For` should have been set with `185.153.176.160` instead of `185.153.176.160, 127.0.0.1`? Maybe Home Assistant does not parse multiple entries "properly" and that's why `header_up` works.

**Comment by francislavoie** [comment]
Right, you're doing trusted_proxies wrong. What you're trusting is the upstream proxy, which is Caddy, which is coming from 127.0.0.1, apparently. So you need `127.0.0.1` in your list in HA config, so that it matches the `127.0.0.1` inside the XFF header and skips it, to get the actual IP which is to the left of that.

Edit: Actually I think the 127.0.0.1 is coming from cloudflared, so you need to trust that anyway cause that's one of the proxies in the chain (user -> cloudflared -> caddy -> HA) so you probably need both Caddy's IP (I guess 172.x.x.x Docker IP) and cloudflared IP (127.0.0.1) in there.

**Comment by felipecrs** [comment]
That was it! It works with:

```yaml
http:
  use_x_forwarded_for: true
  trusted_proxies:
    - 172.30.33.0/24
    - 127.0.0.1
```

And:

```yaml
{
        servers {
                trusted_proxies static 127.0.0.1
                client_ip_headers CF-Connecting-IP
        }
}
```

But I'm not sure that's what I want... passing `127.0.0.1` down is of no use, it does not allow identification of host, and in my case 127.0.0.1 would be always part of `X-Forwarded-For` no matter what.

Maybe some new directive in Caddy to allow omitting certain addresses from `X-Forwarded-For` would be nice.

**Comment by francislavoie** [comment]
> But I'm not sure that's what I want

It is what you want, because that's how trusted proxies work. You need to trust all proxies in the chain, and 127.0.0.1 is one of the proxies in the chain (cloudflared). Here's a long article which explains https://adam-p.ca/blog/2022/03/x-forwarded-for/

**Comment by felipecrs** [comment]
I get it.

But the internal Caddy proxy within the Cloudflared container is supposed to be fully transparent.

And furthermore, I don't want to push this extra burden of changing the HA configuration for every user of the add-on.

I wonder why the equivalent Nginx config works out of the box (I wonder if it always rewrites all hosts in `X-Forwarded-For` or only the `set_real_ip_from` one).

**Comment by francislavoie** [comment]
> But the internal Caddy proxy within the Cloudflared container is supposed to be fully transparent.

That's impossible. TCP is not a transparent protocol. It carries an IP address on the TCP packets.

> And furthermore, I don't want to push this extra burden of changing the HA configuration for every user of the add-on.

It's necessary for security.

> I wonder why the equivalent Nginx config works out of the box (I wonder if it always rewrites all hosts in X-Forwarded-For or only the set_real_ip_from one).

Yes, Nginx does not do it securely, doesn't actually validate the incoming proxy, so it's possible to spoof the IP address if something hits Nginx directly with Cf-Connecting-Ip or whatever.

**Comment by felipecrs** [comment]
> That's impossible. TCP is not a transparent protocol. It carries an IP address on the TCP packets.

Sorry, I meant as transparent as possible.

> Yes, Nginx does not do it securely, doesn't actually validate the incoming proxy, so it's possible to spoof the IP address if something hits Nginx directly with Cf-Connecting-Ip or whatever.

That is why my intention is to only omit `127.0.0.1`, not anything else. If I am already trusting `127.0.0.1` in my Caddyfile, I don't see what the use is for passing it down to Home Assistant and having to make HA trust it too.

If it gets spoofed with any other address, it should be kept as part of `X-Forwarded-For`, so that my Home Assistant can catch it.

**Comment by francislavoie** [comment]
Because like I said, that's how trust works. Everything has to trust everything. There can't be incomplete/partial trust. XFF has every proxy add itself in the header to say "yeah I was in the chain and this is what I validated", but HA still has to "mistrust" the proxy by default in case it was misconfigured or is straight up lying or whatever, so you adding 127.0.0.1 in your trusted_proxies list is saying "yeah HA, don't worry, I validated that the proxy at 127.0.0.1 is chill, I configured it". Read the article I linked, please.

**Comment by felipecrs** [comment]
Alright. I haven't read the article yet but I promise I'll do. I can't promise I'll change my mind about this particular case though.

Once again thank you SO MUCH for the time you spent with me.

Our discussion may help other people in the future when they migrate from Nginx to Caddy.
